### ISSUE TITLE
Async guards should have the option to wait for previous guards to finish

### ISSUE THREAD

**Issue by MickL** [angular/angular#40577]
This is a copy of #21702 which has been locked due to inactivity. Still it is relevant.

### I'm submitting a...

[ ] Regression
[ ] Bug report
[X] Feature request
[ ] Documentation issue or request
[ ] Support request

### Current behavior

When having multiple async guards (returning promise or observable) for the same path they execute at the same time:

```ts
{
  path: '',
  canActivate: [
        AuthGuard,
        DetectLanguage,
        GetAppData,
  ],
  children: [ ... ]
}
```

### Expected behavior

There should be an option to let them execute synchronous: 

- guard[i+1] should wait for guard[i] to finish
- If guard[i] returns false all following guards should not execute

### What is the motivation / use case for changing the behavior?

The example above has an authentication which needs to be done first before getting the app data. AuthGuard returns a token that needs to be used within GetAppData. Currently GetAppData runs immediatly without having a token.

### Workaround

Kinda ugly and confusing workaround:

```ts
{
      path: '',
      canActivate: [
         AuthGuard,
      ],
      children: [
        {
          path: '',
          canActivate: [
             DetectLanguage,
          ],
          children: [
             {
                path: '',
                canActivate: [
                    GetAppdata,
                ],
                component: MyComponent,
             }
         }
      ]
    }
```

**Comment by atscott** [comment]
Totally reasonable feature request. There are several requests for providing an option to modify how the router handles guards, resolvers, activation, etc. in terms of ordering and such. I don't think it makes sense to add all of these different options as things that can be turned on/off in the `ExtraOptions` like we do for certain features now. Instead, I hope the router can be reworked in the future to be more composable where it provides the basic building blocks (creating a target url, determining the target route, collecting guards, etc) and passes these along to the set of operations that developers can swap in and out. This would require quite a bit more design and thinking about what the Router should look like long term.

**Comment by sod** [comment]
You can use the `SequentialRoutingGuardService` from https://ng-run.com/edit/HmMVqtmmHB52knus3yjk?open=app%2Fsequential-routing-guard.service.ts from https://github.com/angular/angular/issues/36538 to achieve this.

**Comment by atscott** [comment]
Closing in favor of #42953 (which links to this for tracking and discoverability). This is a good feature request and the ability to configure this type of thing without a bundle size cost to every Router user even if they don't use it would be one of the main goals there.

**Comment by atscott** [comment]
This will be somewhat trivial in the user-space if/when the ["guards as functions"](https://github.com/angular/angular/pull/46684) change is made.

```
function runSerially(guards: CanActivateFn[]|CanActivateChildFn[]): CanActivateFn|
    CanActivateChildFn {
      return (route: ActivatedRouteSnapshot, state: RouterStateSnapshot) => {
        const injector = coreInject(EnvironmentInjector);
        const observables = guards.map(guard => {
          const guardResult = injector.runInContext(() => guard(route, state));
          return wrapIntoObservable(guardResult).pipe(first());
        })
        return concat(...observables).pipe(takeWhile(v => v === true), last());
      }
}
```

And your route could be 
```
{
  path: '**',
  component: SomeCmp,
  canActivate: [runSerially([guard1, guard2, guard3, guard4])]
}
```
